go/vt/vtgate/engine/concatenate.go

```diff
@@ -285,36 +285,35 @@ func (c *Concatenate) parallelStreamExec(inCtx context.Context, vcursor VCursor,
 		currIndex, currSource := i, source
 		wg.Go(func() error {
 			err := vcursor.StreamExecutePrimitive(ctx, currSource, bindVars, true, func(resultChunk *sqltypes.Result) error {
-				// Process fields when they arrive; coordinate field agreement across sources.
-				if resultChunk.Fields != nil {
-					muFields.Lock()
+				muFields.Lock()
 
+				// Process fields when they arrive; coordinate field agreement across sources.
+				if resultChunk.Fields != nil && rest[currIndex] == nil {
 					// Capture the initial result chunk to determine field types later.
-					if rest[currIndex] == nil {
-						rest[currIndex] = resultChunk
-
-						// If this was the last source to report its fields, derive the final output fields.
-						if !slices.Contains(rest, nil) {
+					rest[currIndex] = resultChunk
+
+					// If this was the last source to report its fields, derive the final output fields.
+					if !slices.Contains(rest, nil) {
+						// We have received fields from all sources. We can now calculate the output types
+						var err error
+						resultChunk.Fields, fieldTypes, err = c.getFieldTypes(vcursor, rest)
+						if err != nil {
 							muFields.Unlock()
-
-							// We have received fields from all sources. We can now calculate the output types
-							var err error
-							resultChunk.Fields, fieldTypes, err = c.getFieldTypes(vcursor, rest)
-							if err != nil {
-								return err
-							}
-
-							defer condFields.Broadcast()
-							return callback(resultChunk, currIndex)
+							return err
 						}
+
+						muFields.Unlock()
+						defer condFields.Broadcast()
+						return callback(resultChunk, currIndex)
 					}
-					// Wait for fields from all sources.
-					for slices.Contains(rest, nil) {
-						condFields.Wait()
-					}
-					muFields.Unlock()
 				}
 
+				// Wait for fields from all sources.
+				for slices.Contains(rest, nil) {
+					condFields.Wait()
+				}
+				muFields.Unlock()
```

Comment:
Why didn't y'all use the `lock ... defer unlock` pattern? That way you call unlock once at line 289 instead of 3 times in various places. And it's future-proof.
Nice work, btw 👍

Reply:
That wouldn't be correct. The lock needs to be released before we call `callback` or we'd deadlock, so using a `defer` doesn't work in 2 of the cases.